### ISSUE TITLE
How to use original micropython machine.SPI methods?

### ISSUE THREAD

**Issue by jd3096-mpy** [lvgl-micropython/lvgl_micropython#193]
It was mentioned in the previous README file.

'All methods that existed for the original machine.SPI are available in the machine.SPI.Device class. They work exactly how they did before.'

I want to use other SPI devices in the lvgl firmware, but I found that the method is not universal.

For example:
`spi = SPI(baudrate=20000000,polarity=0, phase=0,firstbit=SPI.MSB,sck=Pin(LCD_SCK),mosi=Pin(LCD_SI ),miso=Pin(0))`
It can be used in original micropython firmware
`spi = SoftSPI(baudrate=20000000,polarity=0, phase=0,firstbit=SPI.MSB,sck=Pin(LCD_SCK),mosi=Pin(LCD_SI ),miso=Pin(0))` 
can be used but SoftSPI is much slower than hardware SPI

**Comment by kdschlosser** [comment]
All of the methods that were available in the original SPI class are available in the SPI.Device class... You don't have to pass a pin object to the Bus or the Device classes when creating the instances.

so in your case what you would do is this..

```python
import machine

spi_bus = machine.SPI.Bus(
    host=1,
    miso=0,  # Not sure if the 0 is intentional. If you do not want to use this pin set it to -1 not 0. 
    mosi=LCD_SI,
    sck=LCD_SCK
)

device = machine.SPI.Device(
    spi_bus=bus,
    cs=-1,
    freq=20000000,
    phase=0,
    firstbit=SPI.Device.MSB,
    polarity=0,
)


buf = device.read(10)
print(buf)

buf = bytearray(10)
mv = memoryview(buf)
device.readinto(mv)
print(buf)

buf = bytearray(b'this is some data')
mv = memoryview(buf)
device.write(mv)

write_buf = bytearray(b'this is some data')
write_mv = memoryview(write_buf )

read_buf = bytearray(10)
read_mv = memoryview(read_buf)
device.write_readinto(write_mv, read_mv)

print(read_buf)
```


and if you have 2 devices connected and they are running at different speeds you don't have to reinitialize the bus over and over again each time you change what device you want to read/write to. You don't have to manage the cs pin either. It is all handled internally. pass the pin numbers for the cs pins to the SPI.Device class, and pass the appropriate speed. create a dive for each device on your bus and send whatever needs to be sent or receive whatever needs to be collected for data from each of the devices you create. 

The methods are all there and they have not changed. They got moved to the device class but they function in the same manner as they did before.

**Comment by andresponte** [comment]
> All of the methods that were available in the original SPI class are available in the SPI.Device class... You don't have to pass a pin object to the Bus or the Device classes when creating the instances.
> 
> so in your case what you would do is this..
> 
> import machine
> 
> spi_bus = machine.SPI.Bus(
>     host=1,
>     miso=0,  # Not sure if the 0 is intentional. If you do not want to use this pin set it to -1 not 0. 
>     mosi=LCD_SI,
>     sck=LCD_SCK
> )
> 
> device = machine.SPI.Device(
>     spi_bus=bus,
>     cs=-1,
>     freq=20000000,
>     phase=0,
>     firstbit=SPI.Device.MSB,
>     polarity=0,
> )
> 
> 
> buf = device.read(10)
> print(buf)
> 
> buf = bytearray(10)
> mv = memoryview(buf)
> device.readinto(mv)
> print(buf)
> 
> buf = bytearray(b'this is some data')
> mv = memoryview(buf)
> device.write(mv)
> 
> write_buf = bytearray(b'this is some data')
> write_mv = memoryview(write_buf )
> 
> read_buf = bytearray(10)
> read_mv = memoryview(read_buf)
> device.write_readinto(write_mv, read_mv)
> 
> print(read_buf)
> and if you have 2 devices connected and they are running at different speeds you don't have to reinitialize the bus over and over again each time you change what device you want to read/write to. You don't have to manage the cs pin either. It is all handled internally. pass the pin numbers for the cs pins to the SPI.Device class, and pass the appropriate speed. create a dive for each device on your bus and send whatever needs to be sent or receive whatever needs to be collected for data from each of the devices you create.
> 
> The methods are all there and they have not changed. They got moved to the device class but they function in the same manner as they did before.

Hi @kdschlosser,

I'm using a ESP32-C6-LCD-1.47 board and my idea is to communicate with the SD card witch is in the same SPI interface.
I built my bin file with the command:
`python3 make.py esp32 clean --flash-size=4 --enable-jtag-repl=y BOARD=ESP32_GENERIC_C6 DISPLAY=st7789`

I'm able to print the "hello world" message in the display by using the following code:
`import lcd_bus
from micropython import const
import machine


# display settings
_WIDTH = const(172)
_HEIGHT = const(320)
_BL = const(22)
_RST = const(21)
_DC = const(15)

_MOSI = const(6)
_MISO = const(5)
_SCK = const(7)
_HOST = const(1)  # SPI2

_LCD_CS = const(14)
_LCD_FREQ = const(80000000)

_OFFSET_X = const(34)
_OFFSET_Y = const(0)

spi_bus = machine.SPI.Bus(
    host=_HOST,
    mosi=_MOSI,
    miso=_MISO,
    sck=_SCK
)

display_bus = lcd_bus.SPIBus(
    spi_bus=spi_bus,
    freq=_LCD_FREQ,
    dc=_DC,
    cs=_LCD_CS,
)

# we are going to let the display driver sort out the best freame buffer size and where to allocate it to.
# fb1 = display_bus.allocate_framebuffer(_BUFFER_SIZE, lcd_bus.MEMORY_INTERNAL | lcd_bus.MEMORY_DMA)
# fb2 = display_bus.allocate_framebuffer(_BUFFER_SIZE, lcd_bus.MEMORY_INTERNAL | lcd_bus.MEMORY_DMA)

import st7789  # NOQA
import lvgl as lv  # NOQA


display = st7789.ST7789(
    data_bus=display_bus,
    display_width=_WIDTH,
    display_height=_HEIGHT,
    backlight_pin=_BL,
    reset_pin=_RST,
    reset_state=st7789.STATE_LOW,
    backlight_on_state=st7789.STATE_PWM,
    color_space=lv.COLOR_FORMAT.RGB565,
    color_byte_order=st7789.BYTE_ORDER_RGB,
    rgb565_byte_swap=True,
    offset_x=_OFFSET_X,
    offset_y=_OFFSET_Y,
)

import task_handler  # NOQA

display.set_power(True)
display.init()
display.set_backlight(100)

th = task_handler.TaskHandler()

scrn = lv.screen_active()
scrn.set_style_bg_color(lv.color_hex(0x000000), 0)

label = lv.label(scrn)
label.set_text('HELLO WORLD!')
label.align(lv.ALIGN.CENTER, 0, 0)`

I tried to run your code (with the proper MISO/MOSI/CLK/SS pins) but it blocks in the device.read(10) line, if remove it then it blocks in the line device.readinto(mv).

**Comment by kdschlosser** [comment]
I am not seeing any code in there for the SPI driver.

**Comment by andresponte** [comment]
> I am not seeing any code in there for the SPI SD Card driver.
> 

that's correct, I'm not using the driver yet. I just wanted to start from the basics, send a low level SPI command and check the reply... In case there is no answer, I should at least receive 0xFF...but the code blocks when I call any of the SPI functions like write, read... So, I just ran your code, and only that code, without the screen stuff, just to see if I can control the SPI, but it still blocks in any of the SPI functions.

By the way, I shared the screen code just to show at least with the screen driver, the SPI works fine in my micropython bin.

**Comment by kdschlosser** [comment]
are you pulling the CS pin for the SD Card reader low before sending commands to it?

**Comment by kdschlosser** [comment]
actually you cannot even do that because the entire SPI is now managed. If the display is flushing at the same time you pill the SD Card reader pin low you can end up with some funky stuff happening.

Set up the SDCard driver and let er rip. I know it will work without issue. Make sure to pass the same spi bus instance to the SDCard driver that was passed to the lcd bus driver for the display.

**Comment by andresponte** [comment]
> actually you cannot even do that because the entire SPI is now managed. If the display is flushing at the same time you pill the SD Card reader pin low you can end up with some funky stuff happening.
> 
> Set up the SDCard driver and let er rip. I know it will work without issue. Make sure to pass the same spi bus instance to the SDCard driver that was passed to the lcd bus driver for the display.

Hi @kdschlosser, thanks for the repply.
Is there a ESP32-C6 SDCard driver adapted for this managed SPI Bus? I tried "**from machine import SDCard**" but I get "ImportError: can't import name SDCard"

The driver I have is the [sdcard.py](https://github.com/micropython/micropython-lib/blob/master/micropython/drivers/storage/sdcard/sdcard.py), but it uses the micropython normal SPI.

**Comment by kdschlosser** [comment]
so the one built into MicroPython is not available

**Comment by andresponte** [comment]
> so the one built into MicroPython is not available

Correct, it not available in my lvgl micropython build, I don't know why.

**Comment by kdschlosser** [comment]
I can fix that and make it available. It's probably like that because of how the SDCard module is written in MicroPython which combines 2 different SDCard drivers into a single driver.

**Comment by andresponte** [comment]
> I can fix that and make it available. It's probably like that because of how the SDCard module is written in MicroPython which combines 2 different SDCard drivers into a single driver.
> 
Thanks @kdschlosser , should I open a new issue? It seems to be a different topic

**Comment by kdschlosser** [comment]
It is actually fixed in the newest release of MicroPython. I just don't want to upgrade just yet because there are typically a lot of bugs when they do a release so I would rather wait and see what kind of problems pop up with other before I upgrade the MicroPython version.

**Comment by andresponte** [comment]
> It is actually fixed in the newest release of MicroPython. I just don't want to upgrade just yet because there are typically a lot of bugs when they do a release so I would rather wait and see what kind of problems pop up with other before I upgrade the MicroPython version. 

It makes sense! I'll wait then. Thanks for the support.

**Comment by kdschlosser** [comment]
yeah no worries. I might fix it anywhay because I am using a custom SPI driver that I wrote. I will need to add in the bits that remove the MMC Reader portion of the code when running your MCU.
